### PR TITLE
refactor(scheduler): rewrite pipeline scheduler

### PR DESCRIPTION
**Pull request by mwiacx**
## Motivation

Rewrite pipeline scheduler codes to make it more readable，meanwhile we add a new feature for interleaved pipeline scheduler: communication overlap.

## Modification

* rewrite PipelineScheduler and InterleavedPipelineScheduler
* add communication overlap for InterleavedPipelineScheduler

## Use cases

In config:
* `parallel["pipeline"]["size"]` == `1`: Non-PipelineScheduler is used.
* `parallel["pipeline"]["size"]` > `1` and `model["num_chunks"]` == `1`: PipelineScheduler is used.
* `parallel["pipeline"]["size"]` > `1` and `model["num_chunks"]` > `1` and `parallel["pipeline"]["interleaved_overlap"]` == `True`: InterleavedPipelineScheduler with communication overlap is used.
* `parallel["pipeline"]["size"]` > `1` and `model["num_chunks"]` >= `1` and `parallel["pipeline"]["interleaved_overlap"]` == `False`: InterleavedPipelineScheduler without communication overlap is used. 

## Checklist

**Before PR**:

- [x] Pre-commit or other linting tools are used to fix the potential lint issues.
- [x] Bug fixes are fully covered by unit tests, the case that causes the bug should be added in the unit tests.
- [x] The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
- [x] The documentation has been modified accordingly, like docstring or example tutorials.

**After PR**:

- [x] If the modification has potential influence on downstream or other related projects, this PR should be tested with those projects.
- [x] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

File: internlm/core/scheduler/no_pipeline_scheduler.py [hunk 1]
```diff
@@ -68,7 +79,7 @@ def _load_accum_batch(self, data: Any, label: Any):
             label (Any): The label to be loaded.
         """
 
-        _data, _label = self._load_micro_batch(
+        _data, _label = self.load_micro_batch(
```
Comment (huangting4201) [inline internlm/core/scheduler/no_pipeline_scheduler.py:82]: 这里应该是_load_micro_batch？

File: internlm/core/scheduler/no_pipeline_scheduler.py [hunk 2]
```diff
@@ -35,7 +35,12 @@ def data_process_func(dataloader_output):
...
-    def __init__(self, data_process_func: Callable = None, gradient_accumulation_size: int = 1):
+    def __init__(
+        self,
+        data_process_func: Callable = None,
+        gradient_accumulation_size: int = 1,
+        scheduler_hooks: Optional[List[SchedulerHook]] = None,
+    ):
         # check that non-pipeline schedule data process func only takes in one parameter
         # which is the batch data
         if data_process_func:
```
Comment (huangting4201) [inline internlm/core/scheduler/no_pipeline_scheduler.py:46]: data_process_func相关的处理代码我在pp删掉了，no pp忘记删了，逊佬帮忙删了吧

File: internlm/core/scheduler/pipeline_scheduler.py [hunk 3]
```diff
@@ -90,82 +99,90 @@ def __init__(
...
     def load_micro_batch(self):
-        mciro_batch_data, micro_batch_label = self._load_micro_batch(
+        micro_batch_data, micro_batch_label = self._load_micro_batch(
             data=self.batch_data, label=self.batch_label, offset=self.microbatch_offset, micro_bsz=self.microbatch_size
         )
+        micro_batch_data["label"] = micro_batch_label
         self.microbatch_offset += self.microbatch_size
 
         # unpack data process
         # TODO by xyt
```
Comment (huangting4201) [inline internlm/core/scheduler/pipeline_scheduler.py:182]: 彤彤把unpack相关的操作放到dataset那边了，这两行注释逊佬帮忙去掉吧

Comment (mwiacx) [reply internlm/core/scheduler/pipeline_scheduler.py:182]: ok

File: internlm/core/scheduler/no_pipeline_scheduler.py (hunk 2 above)
Comment (mwiacx) [reply internlm/core/scheduler/no_pipeline_scheduler.py:46]: ok

File: internlm/core/scheduler/no_pipeline_scheduler.py (hunk 1 above)
Comment (mwiacx) [reply internlm/core/scheduler/no_pipeline_scheduler.py:82]: ok

File: internlm/core/scheduler/pipeline_scheduler.py [hunk 4]
```diff
@@ -272,163 +292,236 @@ def _backward_step(self, engine, input_obj, output_obj, output_obj_grad):
...
+                engine,
+                input_obj,
+                return_tensors,
+                return_output_label=return_output_label,
+                accum_loss=accum_loss,
+            )
+
+            if not gpc.is_last_rank(ParallelMode.PIPELINE) and need_forward_meta:
+                comm.send_obj_meta(output_obj)
+                need_forward_meta = True  # send only once.
```
Comment (huangting4201) [inline internlm/core/scheduler/pipeline_scheduler.py:337]: send only once 的话，need_forward_meta = True这里应该设置成False？

File: internlm/core/scheduler/pipeline_scheduler.py [hunk 5]
```diff
@@ -461,384 +556,546 @@ def __init__(
...
+        self._send_tensor_shape_flags = [tensor_shape is None for _ in range(num_chunks)]
+
+    def _clear_state(self) -> None:
+        self._accum_loss = None
+        self._return_tensors = None
+        self._input_objs = [[] for _ in range(self._num_chunks)]
+        self._output_objs = [[] for _ in range(self._num_chunks)]
+        self._output_obj_grads = [[] for _ in range(self._num_chunks)]
+
+        self._input_obj_shapes = [self._tensor_shape for _ in range(self._num_chunks)]
```
Comment (huangting4201) [inline internlm/core/scheduler/pipeline_scheduler.py:603]: 这里应该是self.tensor_shape？

File: internlm/core/scheduler/pipeline_scheduler.py (hunk 4 above)
Comment (mwiacx) [reply internlm/core/scheduler/pipeline_scheduler.py:337]: 本地改了

File: internlm/core/scheduler/pipeline_scheduler.py (hunk 5 above)
Comment (mwiacx) [reply internlm/core/scheduler/pipeline_scheduler.py:603]: 已经改了

File: internlm/core/scheduler/pipeline_scheduler.py [hunk 6]
```diff
@@ -461,384 +556,546 @@ def __init__(
...
+        self._run_warmup_loop(
+            engine,
+            num_microsteps,
+            num_warmup_microsteps,
+            receive_extra_backward=False,
+            forward_only=True,
+        )
+
+        output, label = pack_return_tensors(self._return_tensors) if len(self._return_tensors) > 0 else (None, None)
+        return output, label, self._accum_loss
```
Comment (huangting4201) [inline internlm/core/scheduler/pipeline_scheduler.py:1025]: 这里好像不用pack tensor和return output,label，因为forward_backward_step里统一做了

File: internlm/core/scheduler/pipeline_scheduler.py [hunk 7]
```diff
@@ -461,384 +556,546 @@ def __init__(
...
 
-        output_obj = self._call_engine(engine.model[model_chunk_id], data)
+        micro_batch_data = self.load_micro_batch(chunk_id)
+        data, label = self._get_data_label_for_current_step(input_obj, micro_batch_data)
+
+        self._call_hooks("before_forward", data)
+        output_obj = self._call_engine(engine.model[chunk_id], data)
+        # Convert output_obj to fp32 when last model chunk of last stage
+        if gpc.is_pipeline_last_stage(ignore_virtual=False) and isinstance(engine.model[chunk_id], NaiveAMPModel):
+            output_obj = engine.model[chunk_id].convert_to_fp32(output_obj)
```
[the remark this answers is not in the document]
Comment (mwiacx) [reply internlm/core/scheduler/pipeline_scheduler.py:648]: 就是之前说的由于num_chunks的影响，不能再实例化NaiveModel的时候设置是否convert，只能再这个判断到时最前面一个chunk的last stage的时候做混合进度的convert

File: internlm/core/scheduler/pipeline_scheduler.py (hunk 6 above)
Comment (mwiacx) [reply internlm/core/scheduler/pipeline_scheduler.py:1025]: ok

File: train.py [hunk 8]
```diff
@@ -95,12 +96,25 @@ def initialize_model():
...
-        output_to_fp32=is_no_pp_or_last_stage(),
-        dtype=gpc.config.model.get("dtype", torch.half),
-        sync_buffer=False,
-    )
+    if isinstance(model, nn.ModuleList):
+        model = nn.ModuleList(
+            [
+                NaiveAMPModel(
+                    model=_m,
+                    output_to_fp32=False,  # 由interleaved pipleline scheduler手动控制
```
Comment (huangting4201) [inline train.py:104]: 这里的中文注释还没改成英文

File: configs/7B_sft.py [hunk 9]
```diff
@@ -111,6 +111,7 @@
     norm_type="rmsnorm",
     layer_norm_epsilon=1e-5,
     use_flash_attn=True,
+    num_chunks=1,  # if num_chunks > 1, interleaved pipeline scheudler is used.
```
Comment (huangting4201) [inline configs/7B_sft.py:114]: scheudler 拼写错了

File: train.py (hunk 8 above)
Comment (mwiacx) [reply train.py]: 改了

File: configs/7B_sft.py (hunk 9 above)
Comment (mwiacx) [reply configs/7B_sft.py]: 改了

**Review by sunpengsdu — APPROVED** [review]
